Fail to build with Expressive Code disabled

### What version of `starlight` are you using?

0.21.1

### What version of `astro` are you using?

4.5.6

### What package manager are you using?

pnpm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

[Starting](https://starlight.astro.build/getting-started/#create-a-new-project) a new project using `pnpm create astro --template starlight`, [disabling Expressive Code](https://starlight.astro.build/reference/configuration/#expressivecode) and trying to build fails with the following error:

```
11:55:13 [ERROR] [vite] x Build failed in 11.76s
[vite]: Rollup failed to resolve import "virtual:astro-expressive-code/config" from "/home/projects/github-gcerwr/node_modules/astro-expressive-code/components/renderer.ts".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
  Stack trace:
    at viteWarn (/home/projects/github-gcerwr/node_modules/vite/dist/node/chunks/dep-jvB8WLp9.js:67239:27)
    at onwarn (/home/projects/github-gcerwr/node_modules/vite/dist/node/chunks/dep-jvB8WLp9.js:66971:13)
    at Object.logger (/home/projects/github-gcerwr/node_modules/rollup/dist/es/shared/node-entry.js:19975:9)
    at ModuleLoader.resolveDynamicImport (/home/projects/github-gcerwr/node_modules/rollup/dist/es/shared/node-entry.js:18976:58)
```

The only change to the starter template is the edited config to add:

```diff
export default defineConfig({
  integrations: [
    starlight({
      // Disabling Expressive Code for the repro
+      expressiveCode: false,
      title: 'My Docs',
    }),
  ],
});
```

The started template does not involve the `<Code>` component.

cc @hippotastic 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-gcerwr?file=astro.config.mjs

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/rare-timers-judge.md
+++ b/.changeset/rare-timers-judge.md
@@ -1,0 +1,5 @@
+---
+'@astrojs/starlight': patch
+---
+
+Fixes a bundling issue that caused imports from `@astrojs/starlight/components` to fail when using the config setting `expressiveCode: false`.

--- a/packages/starlight/integrations/expressive-code/index.ts
+++ b/packages/starlight/integrations/expressive-code/index.ts
@@ -163,7 +163,40 @@ export const starlightExpressiveCode = ({
 	starlightConfig,
 	useTranslations,
 }: StarlightEcIntegrationOptions): AstroIntegration[] => {
-	if (starlightConfig.expressiveCode === false) return [];
+	// If Expressive Code is disabled, add a shim to prevent build errors and provide
+	// a helpful error message in case the user tries to use the `<Code>` component
+	if (starlightConfig.expressiveCode === false) {
+		const modules: Record<string, string> = {
+			'virtual:astro-expressive-code/api': 'export default {}',
+			'virtual:astro-expressive-code/config': 'export default {}',
+			'virtual:astro-expressive-code/preprocess-config': `throw new Error("Starlight's
+				Code component requires Expressive Code, which is disabled in your Starlight config.
+				Please remove \`expressiveCode: false\` from your config or import Astro's built-in
+				Code component from 'astro:components' instead.")`.replace(/\s+/g, ' '),
+		};
+
+		return [
+			{
+				name: 'astro-expressive-code-shim',
+				hooks: {
+					'astro:config:setup': async ({ updateConfig }) => {
+						updateConfig({
+							vite: {
+								plugins: [
+									{
+										name: 'vite-plugin-astro-expressive-code-shim',
+										enforce: 'post',
+										resolveId: (id) => (id in modules ? `\0${id}` : undefined),
+										load: (id) => (id?.[0] === '\0' ? modules[id.slice(1)] : undefined),
+									},
+								],
+							},
+						});
+					},
+				},
+			},
+		];
+	}
 
 	const configArgs =
 		typeof starlightConfig.expressiveCode === 'object'

--- a/packages/starlight/integrations/expressive-code/index.ts
+++ b/packages/starlight/integrations/expressive-code/index.ts
@@ -179,7 +179,7 @@ export const starlightExpressiveCode = ({
 			{
 				name: 'astro-expressive-code-shim',
 				hooks: {
-					'astro:config:setup': async ({ updateConfig }) => {
+					'astro:config:setup': ({ updateConfig }) => {
 						updateConfig({
 							vite: {
 								plugins: [

